File: datawire/agent.py
# Copyright (C) k736, inc. All Rights Reserved.
# Unauthorized copying or redistribution of this file is strictly prohibited.

import os, sys, time
try:
    import resource
except:
    class resource:
        RUSAGE_SELF = object()
        @staticmethod
        def getrusage(what):
            return object()

from proton import Message, timestamp
from proton.handlers import CHandshaker
from .counts import lib
from .sampler import Sampler
from datawire.stats import app, lib

class SlidingRate:

    def __init__(self, period = 10):
        self.period = period
        self.samples = [(0, 0.0)]*self.period

    def rate(self, count, tstamp):
        self.samples.pop(0)
        self.samples.append((count, tstamp))
        fcount, ftime = self.samples[0]
        return (count - fcount)/(tstamp - ftime)

class Agent:

    def __init__(self, tether, delegate=None):
        self.tether = tether
        if delegate is None:
            self.__delegate = self
        else:
            self.__delegate = delegate
        self.pid = os.getpid()
        self.incoming = SlidingRate()
        self.outgoing = SlidingRate()
        self.incoming_lib = SlidingRate()
        self.outgoing_lib = SlidingRate()
        self.sampler = Sampler(self)
        self.message = Message()
        self.handlers = [CHandshaker(), self.sampler]

    def on_link_local_open(self, event):
        event.link.counts = lib

    def stats(self):
        tstamp = time.time()
        rusage = {}
        ru = resource.getrusage(resource.RUSAGE_SELF)
        for attr in (u'ru_idrss', u'ru_inblock', u'ru_isrss', u'ru_ixrss',
                     u'ru_majflt', u'ru_maxrss', u'ru_minflt', u'ru_msgrcv',
                     u'ru_msgsnd', u'ru_nivcsw', u'ru_nsignals', u'ru_nswap',
                     u'ru_nvcsw', u'ru_oublock', u'ru_stime', u'ru_utime'):
            if hasattr(ru, attr):
                rusage[attr] = getattr(ru, attr)
        result = {u"address": self.tether.address,
                  u"agent": self.tether.agent,
                  u"type": self.tether.agent_type,
                  u"timestamp": timestamp(tstamp*1000),
                  u"pid": self.pid,
                  u"rusage": rusage,
                  u"times": os.times(),
                  u"command": map(unicode, [sys.executable] + sys.argv),
                  u"incoming_count": app.incoming,
                  u"outgoing_count": app.outgoing,
                  u"incoming_count_lib": lib.incoming,
                  u"outgoing_count_lib": lib.outgoing,
                  u"incoming_rate": self.incoming.rate(app.incoming, tstamp),
                  u"outgoing_rate": self.outgoing.rate(app.outgoing, tstamp),
                  u"incoming_rate_lib": self.incoming_lib.rate(lib.incoming, tstamp),
                  u"outgoing_rate_lib": self.outgoing_lib.rate(lib.outgoing, tstamp)}
        return result

    def on_sample(self, event):
        stats = self.stats()
        self.__delegate.sample(stats)
        self.message.body = stats
        event.link.send(self.message).settle()

    def sample(self, stats):
        pass

File: datawire-j/src/main/java/io/datawire/Agent.java
package io.datawire;

import java.util.HashMap;
import java.util.Map;

import org.apache.qpid.proton.amqp.messaging.AmqpValue;
import org.apache.qpid.proton.amqp.messaging.Section;
import org.apache.qpid.proton.engine.Event;
import org.apache.qpid.proton.engine.Handler;
import org.apache.qpid.proton.message.Message;
import org.apache.qpid.proton.reactor.Handshaker;
import org.apache.qpid.proton.reactor.Reactor;

/**
 * Collect and publish statistics about a service.
 * <p>
 * For basic use the application sets up a {@link Tether} and runs a {@link Container} on a {@link Reactor#acceptor(String, int, Handler)}
 * and puts the service handler as container root and the <code>Agent</code> 
 * at address {@link Tether#getAgent()} with {@link Container#put(String, Handler)}
 *
 */
public class Agent extends BaseDatawireHandler {

    /**
     * Agent exposes its statistics map to the service, see {@link Probe}
     */
    public static interface Stats extends Map<String, Object> {
    }

    /**
     * The service can add its own statistics to the statistics collected by the {@link Agent}
     * @see Agent#Agent(Tether, Probe)
     */
    public static interface Probe {
        public void sample(Stats stats);
    }

    private static final Probe NO_PROBE = new Probe() {
        @Override public void sample(Stats stats) { /* nothing */ }
    };

    private final Tether tether;
    private final Probe delegate;
    private final SlidingRate incoming;
    private final SlidingRate outgoing;
    private final SlidingRate incoming_lib;
    private final SlidingRate outgoing_lib;
    private final Sampler sampler;
    private Message message;

    /**
     * @param tether The tether of the service being monitored
     * @param delegate Service callback when collecting samples
     */
    public Agent(Tether tether, Probe delegate) {
        this.tether = tether;
        this.delegate = delegate != null ? delegate : NO_PROBE;
        this.incoming = new SlidingRate();
        this.outgoing = new SlidingRate();
        this.incoming_lib = new SlidingRate();
        this.outgoing_lib = new SlidingRate();
        this.sampler = new Sampler(this);
        this.message = Message.Factory.create();
        add(new Handshaker());
        add(sampler);
    }

    /**
     * Service can adjust sampling frequency using {@link Sampler#setFrequency(float)}
     * @return The sampler used by this agent
     */
    public Sampler getSampler() {
        return sampler;
    }

    private static class StatsImpl extends HashMap<String,Object> implements Stats {
        public void put(Object... pairs) {
            for (int i = 0; i < pairs.length; ) {
                String key = (String)pairs[i++];
                Object value = i < pairs.length ? pairs[i++] : null;
                put(key, value);
            }
        }
    }

    private Stats sample() {
        StatsImpl stats = new StatsImpl();
        long tstamp = System.currentTimeMillis();
        stats.put(
            "address", tether.getAddress(),
            "agent", tether.getAgent(),
            "type", tether.getAgentType(),
            "timestamp", tstamp,
            "pid", DatawirePlatfomUtils.pid(),
            "rusage", DatawirePlatfomUtils.rusage(),
            "times", DatawirePlatfomUtils.times(),
            "command", DatawirePlatfomUtils.command(),
            "incoming_count", Counts.app.getIncoming(),
            "outgoing_count", Counts.app.getOutgoing(),
            "incoming_count_lib", Counts.lib.getIncoming(),
            "outgoing_count_lib", Counts.lib.getOutgoing(),
            "incoming_rate", incoming.rate(Counts.app.getIncoming(), tstamp),
            "outgoing_rate", outgoing.rate(Counts.app.getOutgoing(), tstamp),
            "incoming_rate_lib", incoming_lib.rate(Counts.lib.getIncoming(), tstamp),
            "outgoing_rate_lib", outgoing_lib.rate(Counts.lib.getOutgoing(), tstamp)
        );
        return stats;
    }

    @Override
    public void onSample(DatawireEvent e) {
        Stats stats = sample();
        delegate.sample(stats);
        message.setBody(new AmqpValue(stats));
        DatawireUtils.send(e.getSender(), message);
    }
}
